### DISCUSSION TITLE
BUILD FAILED (macOS 14.6.1 on arm64 using ruby-build 20240727)

### DISCUSSION THREAD

**Discussion by starwels** [rbenv/ruby-build#2439] [Build failures] [answered]
I am using Mac Sonoma 14.6.1

### Steps to reproduce the behavior

rbenv install <version>

### Logs

Output:

```
ruby-build: using openssl@3 from homebrew
==> Downloading ruby-3.2.1.tar.gz...
-> curl -q -fL -o ruby-3.2.1.tar.gz https://cache.ruby-lang.org/pub/ruby/3.2/ruby-3.2.1.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 19.5M  100 19.5M    0     0  20.6M      0 --:--:-- --:--:-- --:--:-- 20.6M
==> Installing ruby-3.2.1...
ruby-build: using readline from homebrew
ruby-build: using libyaml from homebrew
ruby-build: using gmp from homebrew
-> ./configure "--prefix=$HOME/.rbenv/versions/3.2.1" --with-openssl-dir=/opt/homebrew/opt/openssl@3 --enable-shared --with-readline-dir=/opt/homebrew/opt/readline --with-libyaml-dir=/opt/homebrew/opt/libyaml --with-gmp-dir=/opt/homebrew/opt/gmp --with-ext=openssl,psych,+

BUILD FAILED (macOS 14.6.1 on arm64 using ruby-build 20240530)
```

Build log:

```
cd /var/folders/sc/57ssr2rs46bcnznj6s7h0yqc0000gn/T/ruby-build.20240815165202.30373.mE5ssI
==> Downloading ruby-3.2.1.tar.gz...
-> curl -q -fL -o ruby-3.2.1.tar.gz https://cache.ruby-lang.org/pub/ruby/3.2/ruby-3.2.1.tar.gz
-> tar xzf ruby-3.2.1.tar.gz
cd /var/folders/sc/57ssr2rs46bcnznj6s7h0yqc0000gn/T/ruby-build.20240815165202.30373.mE5ssI/ruby-3.2.1
==> Installing ruby-3.2.1...
-> ./configure "--prefix=$HOME/.rbenv/versions/3.2.1" --with-openssl-dir=/opt/homebrew/opt/openssl@3 --enable-shared --with-readline-dir=/opt/homebrew/opt/readline --with-libyaml-dir=/opt/homebrew/opt/libyaml --with-gmp-dir=/opt/homebrew/opt/gmp --with-ext=openssl,psych,+
checking for ruby... /usr/bin/ruby
tool/config.guess already exists
tool/config.sub already exists
checking build system type... aarch64-apple-darwin23.6.0
checking host system type... aarch64-apple-darwin23.6.0
checking target system type... aarch64-apple-darwin23.6.0
checking for llvm-ar... no
checking for clang++... clang++
checking for llvm-nm... no
checking for llvm-objcopy... no
checking for llvm-objdump... no
checking for llvm-ranlib... no
checking for llvm-strip... no
checking for gcc... clang
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether the compiler supports GNU C... yes
checking whether clang accepts -g... yes
checking for clang option to enable C11 features... none needed
checking whether the compiler supports GNU C++... yes
checking whether clang++ accepts -g... yes
checking for clang++ option to enable C++11 features... none needed
checking how to run the C preprocessor... clang -E
checking for ranlib... ranlib
checking for gar... no
checking for ar... ar
checking for gas... no
checking for as... as
checking for gnm... no
checking for nm... nm
checking for gobjcopy... no
checking for objcopy... no
checking for gobjdump... no
checking for objdump... objdump
checking for gstrip... no
checking for strip... strip
checking for stdio.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for strings.h... yes
checking for sys/stat.h... yes
checking for sys/types.h... yes
checking for unistd.h... yes
checking for wchar.h... yes
checking for minix/config.h... no
checking for vfork.h... no
checking whether it is safe to define __EXTENSIONS__... yes
checking whether _XOPEN_SOURCE should be defined... no
checking if minimum required OS X version is supported... yes
checking if thread-local storage is supported... yes
checking whether the linker is GNU ld... no
checking whether clang -E accepts -o... yes
checking for clang... /usr/bin/clang
checking for real target cpu... arm64
checking arch option... 
checking for ar flags... rcu
checking whether ln -s works... yes
checking whether make sets $(MAKE)... yes
checking for a BSD-compatible install... /opt/homebrew/bin/ginstall -c
checking for a race-free mkdir -p... /opt/homebrew/bin/gmkdir -p
checking for dtrace... dtrace
checking for dot... no
checking for doxygen... no
checking for pkg-config... pkg-config
checking whether it is Android... no
checking for cd using physical directory... cd -P
checking whether CFLAGS is valid... yes
checking whether LDFLAGS is valid... no
configure: error: something wrong with LDFLAGS="-L/opt/homebrew/opt/libpq/lib -L/opt/homebrew/opt/libxslt/lib -L/usr/local/opt/openssl/lib"
external command failed with status 1
```

**Comment by IDCjaja** [comment]
I have the same issue, how to fixed this

**Comment by ladaposamuel** [comment]
Had this issue as well when I tried installing `rbenv install 2.7.0`, but `rbenv install 2.7.2` worked - M1 Max

**Comment by mislav** [comment, marked as answer]
From these lines:
```
checking whether LDFLAGS is valid... no
configure: error: something wrong with LDFLAGS="-L/opt/homebrew/opt/libpq/lib -L/opt/homebrew/opt/libxslt/lib -L/usr/local/opt/openssl/lib"
```
to me it sounds like some of the directories listed in your LDFLAGS environment variable are invalid or missing. Furthermore, it seems to refer to a `/usr/local/opt/openssl` directory, which seems wrong to me because it's neither the location of Homebrew OpenSSL nor the system one. Please retry building Ruby with the LDFLAGS environment variable cleared out:
```
LDFLAGS="" rbenv install 3.2.1
```

**Reply by meabbasmehdi** [reply]
Thanks

**Comment by ecorrons** [comment]
Hey guys, I'm facing a similar issue here... I dunno what else to do. I'm using Apple M1 Pro Sonoma 14.6.1
trying to native install Ruby 2.6.10, and I'm not able to do that. Already installed openssl 1.1 compiling it manually. To give you more context of my env:

`$ which openssl`
`/opt/openssl@1.1/bin/openssl`

`rbenv --version`
`rbenv 1.3.0`

`brew --prefix readline`
`/usr/local/opt/readline`

I have set these on my bash_profile:
export LDFLAGS="-L/opt/openssl@1.1/lib"
export CPPFLAGS="-I/opt/openssl@1.1/include"
export PKG_CONFIG_PATH="/opt/openssl@1.1/lib/pkgconfig"
export RUBY_CONFIGURE_OPTS="--with-readline-dir=$(brew --prefix readline) --with-openssl-dir=/opt/openssl@1.1"

So, when I run:

`rbenv install 2.6.10`

```
==> Downloading ruby-2.6.10.tar.bz2...
-> curl -q -fL -o ruby-2.6.10.tar.bz2 https://cache.ruby-lang.org/pub/ruby/2.6/ruby-2.6.10.tar.bz2
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 13.4M  100 13.4M    0     0  10.2M      0  0:00:01  0:00:01 --:--:-- 10.2M
==> Installing ruby-2.6.10...

WARNING: ruby-2.6.10 is past its end of life and is now unsupported.
It no longer receives bug fixes or critical security updates.

ruby-build: using libyaml from homebrew
-> ./configure "--prefix=$HOME/.rbenv/versions/2.6.10" --enable-shared --with-libyaml-dir=/usr/local/opt/libyaml --with-ext=openssl,psych,+ --with-readline-dir=/usr/local/opt/readline --with-openssl-dir=/opt/openssl@1.1
-> make -j 10

BUILD FAILED (macOS 14.6.1 on arm64 using ruby-build 20241213)
```

The log is very long, I've search for an error (there are lots of warnings) and I only found this:

```
../.././include/ruby/ruby.h:1826:24: note: expanded from macro 'rb_intern'
        __extension__ (RUBY_CONST_ID_CACHE((ID), (str))) : \
                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../.././include/ruby/ruby.h:1811:5: note: expanded from macro 'RUBY_CONST_ID_CACHE'
    {                                                   \
    ^
date_core.c:2388:8: warning: '}' and ')' tokens terminating statement expression appear in different macro expansion contexts [-Wcompound-token-split-by-macro]
                vn = f_round(vs);
                     ^~~~~~~~~~~
date_core.c:44:34: note: expanded from macro 'f_round'
#define f_round(x) rb_funcall(x, rb_intern("round"), 0)
                                 ^~~~~~~~~~~~~~~~~~
../.././include/ruby/ruby.h:1826:24: note: expanded from macro 'rb_intern'
        __extension__ (RUBY_CONST_ID_CACHE((ID), (str))) : \
                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../.././include/ruby/ruby.h:1816:5: note: expanded from macro 'RUBY_CONST_ID_CACHE'
    }
    ^
date_core.c:2388:8: note: ')' token is here
                vn = f_round(vs);
                     ^~~~~~~~~~~
date_core.c:44:34: note: expanded from macro 'f_round'
#define f_round(x) rb_funcall(x, rb_intern("round"), 0)
                                 ^~~~~~~~~~~~~~~~~~
../.././include/ruby/ruby.h:1826:56: note: expanded from macro 'rb_intern'
        __extension__ (RUBY_CONST_ID_CACHE((ID), (str))) : \
                                                       ^
1 error generated.
```

Could you please give me a clue of what can I try?

Thanks in advance. Emilia.

**Reply by mislav** [reply]
Hi, yours could have been a new discussion topic, since you are building a very different Ruby version than those mentioned in this thread, and nothing in your build log resembles the problems that folks in this thread had. But let's see if we can get you helped either way.

Viewing build logs can be confusing since it's hard to make sense of most of the output, but one thing to keep in mind is that `note:` and `warning:` messages from a compiler are (unfortunately) common and they are not fatal, so none of them could have caused the build to fail. The error is likely somewhere else in a part of the log that you haven't pasted, but before the "1 error generated" line. You mostly want to search the log file for a line that contains `error:`.

However, I think I see what could be be the problem: you might have migrated from an Intel-based Mac machine to this Apple M1 machine and used Apple's migration assistant to do that. This caught my attention: `brew --prefix readline => /usr/local/opt/readline`. The prefix for Homebrew packages is `/usr/local` on Intel machines, but `/opt/homebrew` for Apple Silicon machines, so when I see `/usr/local` on M1 machine it raises alarm bells for me. The problem with having migrated Homebrew packages this way is that they are still built for the old Intel architecture and Ruby likely won't be able to link to them. The solution would be to reinstall Homebrew and all its current packages—there are readily available resources online for how to do it if you search for it. Then you can installing Ruby 2.6 again.

Some people have also reported in the past that setting the `CFLAGS="-DUSE_FFI_CLOSURE_ALLOC"` environment variable while building Ruby 2.6 can help avoid some compilation problems.

And finally, a word of warning: I see that you've installed OpenSSL 1.1 (which this Ruby version needs) to `/opt/openssl@1.1`, which is fine, but I also see that you've permanently added it to your PATH, which is less fine because it “shadows” the system OpenSSL already present on every Mac and that can cause other software to fail in the future when you're trying to compile it. Best to not add `/opt/openssl@1.1` to your environment variables. In fact, you do not need to set LDFLAGS, CPPFLAGS, PKG_CONFIG_PATH, or RUBY_CONFIGURE_OPTS in your `.bash_profile`, despite what you might have read from other users suggesting on forums including this one. Simply telling the build system where openssl is will do the trick:

```
rbenv install 2.6.10 -- --with-openssl-dir=/opt/openssl@1.1
```
and Ruby itself will use that information to link to it without any other environment variables needed. Additionally, ruby-build will automatically link to readline and libyaml from your Homebrew.

Good luck!

**Reply by ecorrons** [reply]
Hello @mislav , first of all, thank you so much for your reply, what an honor. In the second place, you were so right —I grep better the error line in the log, and then I was able to install Ruby 2.6.10 by reinstalling Homebrew! Now I’m facing other issues trying to install the mysql2 gem.. but that's another story, at least I’ve made some progress and managed to move forward a bit. I’m so grateful :) Merry Christmas to you! Greetings from Argentina.

**Reply by mislav** [reply]
> Now I’m facing other issues trying to install the mysql2 gem..

Ah, sometimes installing gems with native extensions (such as mysql2) requires also passing some build configuration. In the case of mysql2, I think it also needs the `--with-openssl-dir` flag when using a non-standard OpenSSL location. Try this:

```
gem install mysql2 -- --with-openssl-dir=/opt/openssl@1.1
```

or, if installing project gems using Bundler:

```
# this configuration will be remembered for this project:
bundle config build.mysql2 --with-openssl-dir=/opt/openssl@1.1
```

**Reply by ecorrons** [reply]
Worked like a charm, thanks for all your help. Happy New Year!
